autode/ext/src/idpp.cpp

```diff
@@ -0,0 +1,904 @@
+        images.at(0).coords = std::move(init_coords);
+        images.at(n_images - 1).coords = std::move(final_coords);
+    }
+
+    void NEB::fill_linear_interp() {
+        /* Fill the NEB images with a linear interpolation */
+        const auto& coords_0 = images.at(0).coords;
+        const auto& coords_fin = images.at(n_images - 1).coords;
+        for (int k = 1; k < n_images - 1; k++) {
+            double fac = static_cast<double>(k)
+                        / static_cast<double>(n_images - 1);
+            auto new_coords = coords_0 + (coords_fin - coords_0) * fac;
+            images.at(k).coords = new_coords;
+        }
+        reset_k_spr();
+        images_prepared = true;
+    }
+
+    void NEB::fill_sequentially(const IDPPotential& pot,
+                                const int add_maxiter,
+                                const double add_maxgtol) {
+        /* Fill the NEB path sequentially */
+        add_first_two_images();
+        int n_added = 4;
+
+        while (n_added <= n_images) {
+            auto opt = BBMinimiser(add_maxiter, add_maxgtol);
+            auto conv_idx = opt.min_frontier(*this, frontier, pot);
+            if (n_added == n_images) break;
+            add_image_next_to(conv_idx);
+            n_added++;
+        }
+        reset_k_spr();
+        images_prepared = true;
+    }
+
+    double NEB::get_d_id() const {
+        /* Obtain the ideal distance between images at current stage */
+        double dist = 0.0;
+        for (int k = 0; k < frontier.left; k++) {
+            dist += arrx::norm_l2(images[k].coords - images[k+1].coords);
+        }
+        double d_mid = arrx::norm_l2(
+            images[frontier.left].coords - images[frontier.right].coords
+        );
+        dist += d_mid;
+        for (int k = frontier.right; k < n_images-1; k++) {
+            dist += arrx::norm_l2(images[k].coords - images[k+1].coords);
+        }
+        return dist / static_cast<double>(n_images - 1);
+    }
+
+    double NEB::get_k_mid() const {
+        /* Get the force constant between the frontier images */
+        double d_mid = arrx::norm_l2(
+            images[frontier.left].coords - images[frontier.right].coords
+        );
+        return k_spr * get_d_id() / d_mid;
+    }
+
+    void NEB::reset_k_spr() {
+        /* Reset the spring constant to base value for all images */
+        for (int k = 0; k < n_images; k++) images[k].k_spr = k_spr;
+    }
+
+    void NEB::get_engrad(double& en, arrx::array1d& grad) const {
+        /* Obtain the average energy, and gradient of all images excluding the
+         * initial and final images
+         *
+         * Arguments:
+         *   en: (out) Average energy of the NEB
+         *   grad: (out) Array where the NEB gradient will be stored.
+         *          Will be resized if required.
+         */
+        size_t dim = (n_images - 2) * n_atoms * 3;  // required dimension
+        if (grad.size() != dim) grad.resize(dim);
```

Comment:
this _feels_ weird for the function to do. just reading the function signature I wouldn't expect it to modify the array – rather throw if it's not the right size. (same comment applies below)